src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -5688,9 +5721,15 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
     b(LAST_CHECK);
 
     bind(STUB);
-    ldr(tmp4, Address(pre(a2, base_offset)));
-    cmp(cnt2, cnt1);
-    br(NE, DONE);
+    ldr(tmp4, Address(pre(a2, start_offset)));
+    if (is_8aligned) {
+      // Check if lenghts are equal. When bases are
+      // not aligned, we compare the lengths in the
+      // main loop and don't need to compare it
+      // explicitely ahead of the loop.
+      cmp(cnt2, cnt1);
```

Comment:
Why do we need this? Surely if the base isn't required to be aligned, then it might be aligned. So why can't we use the not-aligned version in all cases?

Reply:
The current implementation assumes that the base (first array element) is aligned. In this case, the array length is *not* aligned (8 bytes mark, word, 4 bytes compressed-Klass*, 4 bytes length), that is why in this case we compare the length ahead of the main loop. With uncompressed Klass* (8 bytes mark-word, 8 bytes Klass*, 4 bytes length, ...) or Lilliput (8 bytes mark-word/Klass*, 4 bytes length, ...), the base is only 4-bytes-aligned, but we can start at the length and still enter the main-loop at an 8 bytes aligned address. As a bonus, that also compares the lengths and we can save a few instructions/branches for that.

Reply:
So what we're saying here is not so much that the base is not aligned, but that the length _is_?

Reply:
Yes, exactly. Perhaps makes sense to rename the variable to 'base_is_8aligned' or something similar?

Reply:
> Yes, exactly. Perhaps makes sense to rename the variable to 'base_is_8aligned' or something similar?

With a comment. "Either the base or the length is at an 8-byte-aligned address." This boolean tells you which, and because "reasons" we're trying to maintain alignment. I implore you to measure the difference from alignment, and if the read alignment makes little or no difference, don't do this micro-optimization.

Reply:
This is not about an optimization, but about a correctness issue. The loop(s) have been written under the assumption that they can read full words, which is true if we start at a word boundary. However, if we don't, then we can attempt an (unaligned) read beyond the array, and if that memory is outside of the heap and unmapped, then we would crash. Note that this currently only happens when running with -UseCompressedClassPointers which almost nobody does. We encountered it with Lilliput, which changes array layout in a similar way.

Another way to think of the situation is this: The length and the array elements are a contiguous chunk of memory, starting at the array-length. We want to compare all of it, and fail as soon as we encounter a mismatch. And since the array-length may be at an unaligned location, we compare that 'head' explicitly, and then enter the main-loop at an aligned memory location (the first array element), and can do an optimized loop that reads full words. When running with -CCP (or Lilliput), the first array element will be unaligned, though, but when that happens, we know that the array-length is aligned, and we don't have to compare the unaligned head (length) ahead of the main-loop, but can instead enter the main-loop at the aligned address of the array-length field.

Reply:
> This is not about an optimization, but about a correctness issue. The loop(s) have been written under the assumption that they can read full words, which is true if we start at a word boundary. However, if we don't, then we can attempt an (unaligned) read beyond the array, and if that memory is outside of the heap and unmapped, then we would crash. Note that this currently only happens when running with -UseCompressedClassPointers which almost nobody does. We encountered it with Lilliput, which changes array layout in a similar way.

OK, I see. Fair enough.